lib/routes/shared/security_pin/prompt_pin_code.dart

```diff
@@ -0,0 +1,298 @@
+        padding: EdgeInsets.only(top: 24.0),
+        margin: EdgeInsets.only(bottom: 0),
+        width: _enteredPinCode.length == _passwordLength ? 28 : 24,
+        height: _enteredPinCode.length == _passwordLength ? 28 : 24,
+        decoration: BoxDecoration(
+            color: i < _enteredPinCode.length ? Colors.white : Colors.transparent,
+            shape: BoxShape.circle,
+            border: Border.all(color: Colors.white, width: 2.0)),
+      ));
+    }
+    return list;
+  }
+
+  Widget _numPad() {
+    return new GridView.count(
+        crossAxisCount: 3,
+        childAspectRatio: 5 / 3,
+        padding: EdgeInsets.zero,
+        children: List<int>.generate(9, (i) => i).map((index) => _numberButton((index + 1).toString())).followedBy([
+          Container(
+            child: new IconButton(
+              onPressed: () => _setPinCodeInput(""),
+              icon: Icon(
+                Icons.delete_forever,
+                color: Colors.white,
+              ),
+            ),
+          ),
+          _numberButton("0"),
+          Container(
+            child: new IconButton(
+              onPressed: () => _setPinCodeInput(_enteredPinCode.substring(0, _enteredPinCode.length - 1)),
+              icon: Icon(
+                Icons.backspace,
+                color: Colors.white,
+              ),
+            ),
+          ),
+        ]).toList());
+  }
+
+  Container _numberButton(String number) {
+    return Container(
+      child: new FlatButton(
+        onPressed: () => _onNumButtonPressed(number),
+        child: new Text(number, textAlign: TextAlign.center, style: theme.numPadNumberStyle),
+      ),
+    );
+  }
+
+  _onNumButtonPressed(String numberText) {
+    setState(() {
+      _setErrorMessage("");
+      if (_enteredPinCode.length < _passwordLength) {
+        _setPinCodeInput(_enteredPinCode + numberText);
+      }
+      if (_enteredPinCode.length == _passwordLength) {
+        // Validate current PIN before changing PIN
+        if (((widget.changePassword && _validated) || widget.setPassword) &&  _tmpPassword.isEmpty) {
+          // If a new PIN is being set prompt user to enter the PIN again
+          Future.delayed(Duration(milliseconds: 300), () => _setPinCodeForValidation()); // Wait 300ms for scale animation to end
+        } else {
+          // Check if PINs match if a new PIN is being set or current PIN is being changed
+          Future.delayed(
+              Duration(milliseconds: 300),
+              () => (widget.setPassword || (widget.changePassword && _validated))
+                  ? _matchPinCodes(_enteredPinCode == _tmpPassword)
+                  : _validatePinCode(_enteredPinCode == _user.securityModel.pinCode));
+        }
+      }
+    });
+  }
+
+  _setPinCodeForValidation() {
+    setState(() {
+      _tmpPassword = _enteredPinCode;
+    });
+    _setPinCodeInput("");
+    _setLabel("Re-enter your new PIN");
+  }
+
+  void _validatePinCode(bool isValid) {
```

Comment:
Nit: rename to `_onPinCodeValide(bool isValid)` since it is not really validating, just responding to the validation result that was done earlier.